File: android/src/main/java/com/jimmydaddy/imagemarker/base/MarkTextOptions.java
package com.jimmydaddy.imagemarker.base;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

import java.lang.reflect.Array;

public class MarkTextOptions extends Options {
  public TextOptions[] watermarkTexts;

  public MarkTextOptions(ReadableMap options) {
    super(options);
    ReadableArray waterMarkTextsMap = options.getArray("watermarkTexts");
    if (waterMarkTextsMap.size() <= 0) return;
    watermarkTexts = new TextOptions[waterMarkTextsMap.size()];
    for (int i = 0; i < waterMarkTextsMap.size(); i++) {
      ReadableMap textMap = waterMarkTextsMap.getMap(i);
      watermarkTexts[i] = new TextOptions(textMap);
    }
  }

  static public MarkTextOptions checkParams(ReadableMap opts, Promise promise) {
    try {
      return new MarkTextOptions(opts);
    } catch (MarkerError e) {
      promise.reject(e.getErrorCode(), e.getErrMsg());
    }
    return null;
  }
}

File: android/src/main/java/com/jimmydaddy/imagemarker/base/MarkTextOptions.kt
package com.jimmydaddy.imagemarker.base

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap

class MarkTextOptions(options: ReadableMap) : Options(options) {
  var watermarkTexts: Array<TextOptions>

  init {
    watermarkTexts = readWatermarkTexts(options.getArray("watermarkTexts"))
  }

  companion object {
    internal fun readWatermarkTexts(watermarkTexts: ReadableArray?): Array<TextOptions> {
      if (watermarkTexts == null || watermarkTexts.size() <= 0) {
        throw MarkerError(ErrorCode.PARAMS_REQUIRED, "watermarkTexts is required")
      }

      return Array(watermarkTexts.size()) { index ->
        val textMap = watermarkTexts.getMap(index)
          ?: throw MarkerError(ErrorCode.NULL_MAP, "watermarkTexts[$index] is null")
        TextOptions(textMap)
      }
    }

    @JvmStatic
    fun checkParams(opts: ReadableMap, promise: Promise): MarkTextOptions? {
      try {
        return MarkTextOptions(opts)
      } catch (e: Exception) {
        val markerError = MarkerError.fromInvalidParams(e, "Invalid text marker options")
        promise.reject(markerError.getErrorCode(), markerError.getErrMsg())
      }
      return null
    }
  }
}
